src/Analyzer/Passes/OptimizeRedundantFunctionsInOrderByPass.cpp

```diff
@@ -0,0 +1,120 @@
+#include <Analyzer/Passes/OptimizeRedundantFunctionsInOrderByPass.h>
+#include <Analyzer/ColumnNode.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/QueryNode.h>
+#include <Analyzer/SortNode.h>
+#include <Functions/IFunction.h>
+
+namespace DB
+{
+
+namespace
+{
+
+class OptimizeRedundantFunctionsInOrderByVisitor : public InDepthQueryTreeVisitor<OptimizeRedundantFunctionsInOrderByVisitor>
+{
+
+    struct RedundancyVerdict
+    {
+        bool redundant = true;
+        bool done = false;
+    };
+
+    static constexpr RedundancyVerdict makeNonRedundant() noexcept { return { .redundant = false, .done = true }; }
+
+    std::unordered_set<String> existing_keys;
+
+    RedundancyVerdict isRedundantExpression(FunctionNode * function)
+    {
+        if (function->getArguments().getNodes().empty())
+            return makeNonRedundant();
+
+        if (!function->getFunction()->isDeterministicInScopeOfQuery())
+            return makeNonRedundant();
+
+        // TODO: handle constants here
+        for (auto & arg : function->getArguments().getNodes())
```

Comment:
Do we support only function with single argument ? The loop looks strange, because we iterate single time.

Reply:
It supports functions with multiple arguments. There is also a test for it:
```
EXPLAIN QUERY TREE run_passes=1 SELECT key, a FROM test ORDER BY key, a, exp(key + a);
```